File: web/charts.js
/*
 * charts.js
 * ---------
 * The data marks, drawn as inline SVG.
 *
 * Same rules as the desktop app. Magnitude uses one hue's ramp, so the bars
 * read as a single measure. Identity uses fixed categorical slots, so a
 * category keeps its colour when the data changes. Past eight categories the
 * tail folds into "Other" rather than inventing a ninth hue that nothing could
 * reliably tell apart.
 */

import { byCategory, money, monthlyEquivalent } from "./data.js";

const MAX_BARS = 11;
const MAX_SLOTS = 7;
const BAR_HEIGHT = 15;
const BAR_PITCH = 30;

const SERIES = Array.from({ length: 8 }, (_, i) => `var(--series-${i + 1})`);
// The readable ink for each fill, measured rather than assumed — on the light
// palette several series are dark enough that black text would fail.
const SERIES_INK = Array.from({ length: 8 }, (_, i) => `var(--series-${i + 1}-ink)`);

const svgEl = (name, attrs = {}) => {
  const node = document.createElementNS("http://www.w3.org/2000/svg", name);
  for (const [key, value] of Object.entries(attrs)) node.setAttribute(key, value);
  return node;
};

/** Category totals, with the tail beyond the slot ceiling folded into "Other". */
export function foldedCategories(bills) {
  const shares = byCategory(bills);
  if (shares.length <= MAX_SLOTS + 1) return shares;

  const head = shares.slice(0, MAX_SLOTS);
  const tail = shares.slice(MAX_SLOTS).reduce((sum, [, value]) => sum + value, 0);
  const existing = head.find(([name]) => name === "Other");
  if (existing) existing[1] += tail;
  else head.push(["Other", tail]);
  // Absorbing the tail can push "Other" above a category that outranked it,
  // so re-sort — otherwise the ribbon runs 45%, 15%, 16% and looks broken.
  return head.sort((a, b) => b[1] - a[1]);
}

export function categoryColour(bills, category) {
  const index = foldedCategories(bills).findIndex(([name]) => name === category);
  return index === -1 ? "var(--ink-muted)" : SERIES[index % SERIES.length];
}

/* ------------------------------------------------------------------ bars */

/**
 * One bar per bill, largest first. Returns how many were drawn so the caller
 * can say when it is showing only the top of the list — a silent top-N reads
 * as "this is everything" when it is not.
 */
export function renderBars(host, bills) {
  host.replaceChildren();
  if (!bills.length) {
    host.innerHTML =
      '<p class="empty"><strong>Nothing to chart yet</strong>Add a bill to see the breakdown</p>';
    return 0;
  }

  // Bars compare what each bill costs per month, so a yearly premium sits
  // beside a monthly one honestly instead of dwarfing it twelvefold.
  const shown = [...bills]
    .map((bill) => ({ ...bill, monthly: monthlyEquivalent(bill) }))
    .sort((a, b) => b.monthly - a.monthly)
    .slice(0, MAX_BARS);
  const largest = shown[0].monthly;
  const height = shown.length * BAR_PITCH;

  const svg = svgEl("svg", {
    viewBox: `0 0 1000 ${height}`,
    width: "100%",
    height,
    preserveAspectRatio: "none",
    role: "img",
    "aria-label": `Bills by amount, ${shown.length} shown`,
  });

  const gutter = 190;
  const valueSpace = 130;
  const span = 1000 - gutter - valueSpace;

  shown.forEach((bill, index) => {
    const y = index * BAR_PITCH;
    const share = bill.monthly / largest;
    const width = Math.max(4, span * share);
    // Step a single hue by the value's share of the largest, easing the low
    // end so small bars stay clear of the surface.
    const step = Math.round(Math.pow(share, 0.62) * 8) + 1;

    const label = svgEl("text", {
      x: gutter - 14, y: y + BAR_HEIGHT / 2 + 1,
      "text-anchor": "end", "dominant-baseline": "middle",
      class: "bar-label",
    });
    label.textContent = bill.name;
    svg.append(label);

    svg.append(svgEl("rect", {
      x: gutter, y, width, height: BAR_HEIGHT,
      rx: BAR_HEIGHT / 2, ry: BAR_HEIGHT / 2,
      fill: `var(--ramp-${step})`,
    }));

    const value = svgEl("text", {
      x: gutter + width + 12, y: y + BAR_HEIGHT / 2 + 1,
      "dominant-baseline": "middle", class: "bar-value",
    });
    value.textContent = money(bill.monthly);
    svg.append(value);
  });

  host.append(svg);
  return shown.length;
}

/* ---------------------------------------------------------------- ribbon */

/** A stacked bar of the monthly total, split by category, plus its legend. */
export function renderRibbon(barHost, legendHost, bills) {
  barHost.replaceChildren();
  legendHost.replaceChildren();
  if (!bills.length) return;

  const shares = foldedCategories(bills);
  const grand = shares.reduce((sum, [, value]) => sum + value, 0);
  const gap = 3;

  const svg = svgEl("svg", {
    viewBox: "0 0 1000 30", width: "100%", height: 30,
    preserveAspectRatio: "none", role: "img",
    "aria-label": "Share of monthly bills by category",
  });

  let cursor = 0;
  shares.forEach(([name, amount], index) => {
    const width = (amount / grand) * 1000;
    const x = cursor + (index ? gap / 2 : 0);
    const w = Math.max(2, width - (index ? gap / 2 : 0) - (index < shares.length - 1 ? gap / 2 : 0));
    const first = index === 0;
    const last = index === shares.length - 1;

    svg.append(svgEl("rect", {
      x, y: 0, width: w, height: 30,
      rx: first || last ? 15 : 5, ry: first || last ? 15 : 5,
      fill: SERIES[index % SERIES.length],
    }));

    const percent = amount / grand;
    if (w > 62) {
      const text = svgEl("text", {
        x: x + w / 2, y: 16, "text-anchor": "middle",
        "dominant-baseline": "middle", class: "ribbon-label",
        fill: SERIES_INK[index % SERIES_INK.length],
      });
      text.textContent = `${Math.round(percent * 100)}%`;
      svg.append(text);
    }
    cursor += width;
  });
  barHost.append(svg);

  // The legend is always present, so identity is never carried by colour alone.
  for (const [index, [name, amount]] of shares.entries()) {
    const row = document.createElement("div");
    const swatch = document.createElement("span");
    swatch.className = "dot";
    swatch.style.background = SERIES[index % SERIES.length];
    row.append(swatch, document.createTextNode(name + " "));
    const amt = document.createElement("span");
    amt.className = "amt";
    amt.textContent = "$" + Math.round(amount).toLocaleString("en-US");
    row.append(amt);
    legendHost.append(row);
  }
}

/* ----------------------------------------------------------------- meter */

/**
 * How much of the month's income the bills consume.
 *
 * A single ratio against a limit is a meter, not a chart. When bills exceed
 * income the overspend is drawn past the end of the track, so the bar can
 * never silently cap at 100%.
 */
export function renderMeter(fill, over, incoming, outgoing) {
  if (incoming <= 0) {
    fill.style.width = "0%";
    over.style.display = "none";
    return;
  }
  const span = Math.max(incoming, outgoing);
  fill.style.width = `${(Math.min(outgoing, incoming) / span) * 100}%`;

  if (outgoing > incoming) {
    over.style.display = "block";
    over.style.left = `${(incoming / span) * 100}%`;
    over.style.width = `${((outgoing - incoming) / span) * 100}%`;
  } else {
    over.style.display = "none";
  }
}

/**
 * The words beside the meter, and which tone should colour them. The status
 * colour never carries the meaning alone, so the overspend case is spelled out.
 */
export function meterCaption(incoming, outgoing) {
  if (incoming <= 0) {
    return { text: "Add your income to see what's left over", tone: "muted" };
  }
  const remaining = incoming - outgoing;
  const share = Math.round((outgoing / incoming) * 100);
  if (remaining < 0) {
    return {
      text: `Bills exceed income by ${money(-remaining)} · ${share}% of ${money(incoming)}`,
      tone: "critical",
    };
  }
  return {
    text: `${money(remaining)} left over · bills take ${share}% of ${money(incoming)}`,
    tone: "normal",
  };
}

File: desktop/charts.py
"""
charts.py
---------
The three data marks, drawn directly with Pillow.

Each renderer is a pure function - size, data, palette in; an RGBA image out -
so they can be rendered and inspected without opening a window.

Drawing these by hand rather than through a plotting library is what lets the
marks carry gradients, glow and rounded data-ends while staying honest: colour
is still assigned by the job it does. Magnitude uses one hue's ramp; identity
uses the fixed categorical slots so a category keeps its colour as data
changes; overspend uses the reserved critical colour and always ships with
words beside it.
"""

from __future__ import annotations

from PIL import Image

import paint
from models import Bill, Income, by_category, total
from theme import TYPE, Palette

MAX_BARS = 11
MAX_CATEGORY_SLOTS = 7
BAR_HEIGHT = 15
BAR_GAP = 12
LABEL_GUTTER = 132
VALUE_GUTTER = 92
SEGMENT_GAP = 3
RIBBON_HEIGHT = 30


def _blank(size: tuple[int, int]) -> Image.Image:
    return Image.new("RGBA", size, (0, 0, 0, 0))


def bar_capacity(height: int) -> int:
    """How many bars fit in `height`.

    Exposed so the caller can say when the chart is showing only part of the
    list - a silent top-N reads as "this is everything" when it isn't.
    """
    return max(1, min(MAX_BARS, (height + BAR_GAP) // (BAR_HEIGHT + BAR_GAP)))


def _money(value: float) -> str:
    """Currency with the sign outside the symbol: -$920.73, not $-920.73."""
    return f"-${abs(value):,.2f}" if value < 0 else f"${value:,.2f}"


def _truncate(text: str, face, limit: int) -> str:
    if paint.measure(text, face)[0] <= limit:
        return text
    while text and paint.measure(text + "…", face)[0] > limit:
        text = text[:-1]
    return text + "…"


# ------------------------------------------------------------------- meter


def income_meter(
    size: tuple[int, int], bills: list[Bill], incomes: list[Income], palette: Palette
) -> Image.Image:
    """How much of the month's income the bills consume.

    A single ratio against a limit is a meter, not a chart: one track whose
    full width is income, filled to the share the bills take. Overspend is
    drawn past the track's end in the critical colour, so the bar can never
    silently cap at 100%.
    """
    image = _blank(size)
    width, height = size
    outgoing, incoming = total(bills), total(incomes)
    track_height = 14
    top = (height - track_height) // 2

    if incoming <= 0:
        paint.rounded_rect(image, (0, top, width, top + track_height),
                           track_height // 2, fill=palette.track)
        return image

    span = max(incoming, outgoing)
    scale = width / span
    filled = min(outgoing, incoming) * scale
    overspent = outgoing > incoming

    paint.rounded_rect(image, (0, top, incoming * scale, top + track_height),
                       track_height // 2, fill=palette.track)

    if filled > 2:
        # Glow first, so the filled portion reads as lit rather than painted.
        halo = paint.glow((int(filled), track_height), track_height // 2, 11,
                          palette.accent, 120)
        image.alpha_composite(halo, (-22, top - 22))
        bar = paint.linear_gradient((max(2, int(filled)), track_height),
                                    (palette.accent, palette.accent_soft), horizontal=True)
        image.paste(bar, (0, top),
                    paint.rounded_mask((max(2, int(filled)), track_height),
                                       track_height // 2))

    if overspent:
        start, end = incoming * scale, outgoing * scale
        over = paint.glow((int(end - start), track_height), track_height // 2, 10,
                          palette.critical, 150)
        image.alpha_composite(over, (int(start) - 20, top - 20))
        paint.rounded_rect(image, (start + SEGMENT_GAP, top, end, top + track_height),
                           track_height // 2, fill=paint.rgba(palette.critical))
    return image


def meter_caption(bills: list[Bill], incomes: list[Income]) -> tuple[str, str]:
    """The words beside the meter, and which palette role should colour them.

    The status colour never carries the meaning alone, so the overspend case is
    stated in words here rather than left to the red bar.
    """
    outgoing, incoming = total(bills), total(incomes)
    if incoming <= 0:
        return ("Add your income to see what's left over", "muted")
    remaining = incoming - outgoing
    share = outgoing / incoming
    if remaining < 0:
        return (f"Bills exceed income by {_money(-remaining)} · "
                f"{share:.0%} of {_money(incoming)}", "critical")
    return (f"{_money(remaining)} left over · bills take {share:.0%} "
            f"of {_money(incoming)}", "secondary")


# -------------------------------------------------------------------- bars


def bills_by_amount(
    size: tuple[int, int], bills: list[Bill], palette: Palette
) -> Image.Image:
    """One bar per bill, largest first.

    The job is comparing magnitude, so colour is sequential - a single hue's
    ramp, stepped by each bill's share of the largest - not a different hue per
    bill, which would imply the bills are unrelated categories.
    """
    image = _blank(size)
    width, height = size
    if not bills:
        paint.draw_text(image, (width // 2, height // 2),
                        "Add a bill to see the breakdown", TYPE["body"],
                        palette.ink_muted, anchor="mm")
        return image

    ranked = sorted(bills, key=lambda bill: bill.amount, reverse=True)
    shown = ranked[:bar_capacity(height)]
    largest = shown[0].amount
    span = max(1, width - LABEL_GUTTER - VALUE_GUTTER)

    # Spread the rows across the height available rather than stacking them at
    # a fixed pitch and leaving a slab of dead space beneath.
    pitch = BAR_HEIGHT + BAR_GAP
    if len(shown) > 1:
        pitch = min(BAR_HEIGHT + 30, max(pitch, (height - BAR_HEIGHT) / (len(shown) - 1)))

    name_face, value_face = TYPE["small"], TYPE["figure_small"]
    for index, bill in enumerate(shown):
        y = int(index * pitch)  # pastes take integer boxes
        length = max(3, round(span * (bill.amount / largest)))
        color = _ramp_step(palette.sequential, bill.amount / largest)

        paint.draw_text(image, (LABEL_GUTTER - 14, y + BAR_HEIGHT / 2),
                        _truncate(bill.name, name_face, LABEL_GUTTER - 22),
                        name_face, palette.ink_secondary, anchor="rm")

        if index == 0:
            halo = paint.glow((length, BAR_HEIGHT), BAR_HEIGHT // 2, 10, color, 105)
            image.alpha_composite(halo, (LABEL_GUTTER - 20, y - 20))

        # A gradient along the bar keeps the mark from reading as a flat slab.
        # Only the data end is rounded: the bar stays anchored to its zero.
        bar = paint.linear_gradient(
            (length, BAR_HEIGHT),
            (paint.mix(color, palette.surface, 0.18), color),
            horizontal=True,
        )
        image.paste(bar, (LABEL_GUTTER, y),
                    paint.rounded_mask((length, BAR_HEIGHT), BAR_HEIGHT // 2,
                                       corners=(False, True, True, False)))

        paint.draw_text(image, (LABEL_GUTTER + length + 12, y + BAR_HEIGHT / 2),
                        _money(bill.amount), value_face, palette.ink, anchor="lm")
    return image


def _ramp_step(ramp: tuple[str, ...], share: float) -> str:
    """Pick a ramp step for a value's share of the largest, easing the low end."""
    position = max(0.0, min(1.0, share)) ** 0.62
    return ramp[min(len(ramp) - 1, round(position * (len(ramp) - 1)))]


# ------------------------------------------------------------------ ribbon


def category_ribbon(
    size: tuple[int, int], bills: list[Bill], palette: Palette
) -> Image.Image:
    """A stacked bar of the monthly total, split by category, plus its legend.

    Part-to-whole across named categories, so colour is categorical and drawn
    from fixed slots. A stacked bar rather than a pie: shares stay comparable
    against a common baseline and long category names can be labelled directly.
    """
    image = _blank(size)
    width, height = size
    if not bills:
        return image

    shares = _folded(bills)
    grand_total = sum(shares.values())
    scale = width / grand_total

    entries = []
    cursor = 0.0
    items = list(shares.items())
    for index, (name, amount) in enumerate(items):
        color = palette.series[index % len(palette.series)]
        start = cursor * scale + (SEGMENT_GAP / 2 if index else 0)
        end = (cursor + amount) * scale - (SEGMENT_GAP / 2 if index < len(items) - 1 else 0)
        radius = RIBBON_HEIGHT // 2 if index in (0, len(items) - 1) else 5
        paint.rounded_rect(image, (start, 0, end, RIBBON_HEIGHT), radius,
                           fill=paint.rgba(color))

        share = amount / grand_total
        label = f"{share:.0%}"
        if (end - start) > paint.measure(label, TYPE["small_strong"])[0] + 18:
            paint.draw_text(image, ((start + end) / 2, RIBBON_HEIGHT / 2), label,
                            TYPE["small_strong"], paint.readable_on(color), anchor="mm")
        entries.append((name, amount, color))
        cursor += amount

    _legend(image, entries, palette, top=RIBBON_HEIGHT + 18, width=width)
    return image


def _folded(bills: list[Bill]) -> dict[str, float]:
    """Category totals, with everything past the slot ceiling folded into 'Other'.

    A generated ninth hue would be indistinguishable from an existing one under
    colour-vision deficiency, so the tail is merged rather than given a colour.
    """
    shares = by_category(bills)
    if len(shares) <= MAX_CATEGORY_SLOTS + 1:
        return shares
    head = dict(list(shares.items())[:MAX_CATEGORY_SLOTS])
    head["Other"] = head.get("Other", 0.0) + sum(list(shares.values())[MAX_CATEGORY_SLOTS:])
    # Absorbing the tail can push "Other" above a category that outranked it,
    # so re-sort - otherwise the ribbon runs 45%, 15%, 16% and looks broken.
    return dict(sorted(head.items(), key=lambda kv: kv[1], reverse=True))


def _legend(
    image: Image.Image, entries: list, palette: Palette, top: int, width: int
) -> None:
    """Swatch, name and amount in a grid - identity is never colour alone."""
    columns = 3 if width < 520 else 4
    column_width = width // columns
    name_face, value_face = TYPE["small"], TYPE["figure_small"]

    for index, (name, amount, color) in enumerate(entries):
        x = (index % columns) * column_width
        y = top + (index // columns) * 22
        paint.rounded_rect(image, (x, y + 3, x + 9, y + 12), 3, fill=paint.rgba(color))
        # Amount follows the name directly rather than being flushed to the
        # column edge, so each entry reads as one unit instead of two columns.
        amount_text = f"${amount:,.0f}"
        budget = column_width - 30 - paint.measure(amount_text, value_face)[0]
        label = _truncate(name, name_face, budget)
        paint.draw_text(image, (x + 16, y + 1), label, name_face, palette.ink_secondary)
        paint.draw_text(image, (x + 22 + paint.measure(label, name_face)[0], y + 1),
                        amount_text, value_face, palette.ink_muted)


def legend_rows(bills: list[Bill], width: int) -> int:
    """How many legend rows `category_ribbon` will draw, for height budgeting."""
    if not bills:
        return 0
    columns = 3 if width < 520 else 4
    return (len(_folded(bills)) + columns - 1) // columns
